File: 2.2 Linear DS/1D Array Manipulation/p12356.py
# Competitive Programming 3
# Problem 12356

while True:
    S, B = [int(x) for x in input().split()]
    if S == 0 and B == 0: break
    
    # ignore element 0, just there to fix indexing
    left, right = [x-1 for x in range(S+1)], [x+1 for x in range(S+1)]
    
    left[0], left[-1], right[0], right[-1] = 0, 0, 0, 0
    
    for _ in range(B):
        # range to delete
        L, R = [int(x) for x in input().split()]
        left[right[R]] = left[L]
        right[left[L]] = right[R]
                
        if left[L] == 0: print('*',end=' ')
        else: print(left[L], end=' ')
    
        if right[R] == 0: print('*')
        else: print(right[R])
    
    
    print('-')  

File: 2.2 Linear DS/1D Array Manipulation/p12356.cpp
// Competitive Programming 3
// Problem 12356

#include <stdio.h>

int left[100002], right[100002];

int main() {
    int S, B, L, R;

    while(1){
        scanf("%d %d", &S, &B);
        if (S == 0 && B == 0){
            break;
        }

        for (int i = 1; i < S+1; i++){
            left[i] = i-1;
            right[i] = i+1;
        }

        right[S] = 0;

        for (int i = 0; i < B; i++){
            scanf("%d %d", &L, &R);
            left[right[R]] = left[L];
            right[left[L]] = right[R];

            if (left[L] == 0){
                printf("* ");
            }
            else{
                printf("%d ", left[L]);
            }

            if (right[R] == 0){
                printf("*\n");
            }
            else{
                printf("%d\n", right[R]);
            }

        }

        printf("-\n");
    }


}
